scripts: twister: testinstance: store run id between builds

nordic-piks opened the pull request:
**Motivation:**
Incremental builds (as much as possible) using twister.

**Challenge:**
At every cmake call from the twister, there is added `-DTC_RUNID="some_random_value"`,
which is translated into a compiler flag (for all compiler invocations).
Thus, every compilation is different (even within the same build folder, even when there was no code change!),
as the compiler's parameter changes.
This prevents incremental compilation - the build system must recompile each file when compiler invocation differs.

**Proposed solution:**
~~Keep using the Run ID feature (adding a random number at the end of the test output, which later can be verified with the expected value by twister).
Pass randomly generated value to the build system by the file i.e. store the value in the file and pass the file path to the build system via parameter
(the same as previously the random value was passed).
The test framework (ztest) takes that file path, loads file content at compilation time, and includes the random value in the test output.~~
EDIT:
Store run id value in the file (at build folder). When next twister run is executed, read that run id value from file and reuse.
This will allow keeping the same cmake arguments between runs and perform incremental build.

**Example twister call**
`./zephyr/scripts/twister --ninja --inline-logs  --outdir twister-build  --jobs 1 -T zephyr/tests/net/buf --platform nrf52840dk_nrf52840 -v -v --no-clean --device-testing --device-serial /dev/ttyACM0`

`--no-clean ` is crucial.

At the first run, there are compiled 143 files (looking at build.log).
Second run (no code change), only 14!
Previously, at every twister run, there were all files compiled (i.e. 143)

hakehuang reviewed (dismissed):
looks great, the same approach can be use to optimize the build efficiency see below
https://github.com/zephyrproject-rtos/zephyr/issues/62372#issuecomment-1733383028

nashif commented on scripts/pylib/twister/twisterlib/runner.py:
```diff
@@ -1028,8 +1025,31 @@ def cmake_assemble_args(extra_args, handler, extra_conf_files, extra_overlay_con
+        run_id_folder = self.get_run_id_folder(build_dir)
+        return os.path.join(run_id_folder, 'run_id.h')
+
+    def prepare_run_id_file(self, build_dir, run_id_value):
```
was any of this verified on windows?

nordic-piks replied:
yes, working fine.

nordic-piks replied:
Thanks ;)

nashif commented:
Not a big fan of the injection of header files to do this, did you consider reusing the run id from cache, this works for me and is less intrusive:

```patch
diff --git a/scripts/pylib/twister/twisterlib/testinstance.py b/scripts/pylib/twister/twisterlib/testinstance.py
index 958019b411a..cbf72d8ecfb 100644
--- a/scripts/pylib/twister/twisterlib/testinstance.py
+++ b/scripts/pylib/twister/twisterlib/testinstance.py
@@ -15,6 +15,7 @@ from twisterlib.testsuite import TestCase, TestSuite
 from twisterlib.platform import Platform
 from twisterlib.error import BuildError
 from twisterlib.size_calc import SizeCalculator
+from twisterlib.cmakecache import CMakeCache
 from twisterlib.handlers import (
     Handler,
     SimulationHandler,
@@ -53,7 +54,6 @@ class TestInstance:
         self.retries = 0

         self.name = os.path.join(platform.name, testsuite.name)
-        self.run_id = self._get_run_id()
         self.dut = None
         if testsuite.detailed_test_id:
             self.build_dir = os.path.join(outdir, platform.name, testsuite.name)
@@ -62,6 +62,7 @@ class TestInstance:
             source_dir_rel = testsuite.source_dir_rel.rsplit(os.pardir+os.path.sep, 1)[-1]
             self.build_dir = os.path.join(outdir, platform.name, source_dir_rel, testsuite.name)

+        self.run_id = self._get_run_id()
         self.domains = None

         self.run = False
@@ -85,10 +86,22 @@ class TestInstance:
         """ generate run id from instance unique identifier and a random
         number"""

-        hash_object = hashlib.md5(self.name.encode())
-        random_str = f"{random.getrandbits(64)}".encode()
-        hash_object.update(random_str)
-        return hash_object.hexdigest()
+        cmake_cache_path = os.path.join(self.build_dir, "CMakeCache.txt")
+        try:
+            cache = CMakeCache.from_file(cmake_cache_path)
+            cmake_conf = {}
+            for k in iter(cache):
+                cmake_conf[k.name] = k.value
+            run_id = cmake_conf.get('TC_RUNID')
+            logger.info(f"cached run id: {run_id}")
+        except FileNotFoundError:
+            hash_object = hashlib.md5(self.name.encode())
+            random_str = f"{random.getrandbits(64)}".encode()
+            hash_object.update(random_str)
+            run_id = hash_object.hexdigest()
+            logger.info(f"new run id: {run_id}")
+
+        return run_id

     def add_missing_case_status(self, status, reason=None):
         for case in self.testcases:
diff --git a/subsys/testsuite/ztest/CMakeLists.txt b/subsys/testsuite/ztest/CMakeLists.txt
index 7817bc2f30b..43520c82dc1 100644
--- a/subsys/testsuite/ztest/CMakeLists.txt
+++ b/subsys/testsuite/ztest/CMakeLists.txt
@@ -11,6 +11,7 @@ zephyr_include_directories(
   )

 if(DEFINED TC_RUNID)
+  set(TC_RUNID ${TC_RUNID} CACHE STRING "testsuite run id")
   zephyr_compile_definitions(TC_RUNID=${TC_RUNID})
 endif()
```

nordic-piks commented:
> Not a big fan of the injection of header files to do this, did you consider reusing the run id from cache, this works for me and is less intrusive:
> 
> ```diff
> diff --git a/scripts/pylib/twister/twisterlib/testinstance.py b/scripts/pylib/twister/twisterlib/testinstance.py
> index 958019b411a..cbf72d8ecfb 100644
> --- a/scripts/pylib/twister/twisterlib/testinstance.py
> +++ b/scripts/pylib/twister/twisterlib/testinstance.py
> @@ -15,6 +15,7 @@ from twisterlib.testsuite import TestCase, TestSuite
>  from twisterlib.platform import Platform
>  from twisterlib.error import BuildError
>  from twisterlib.size_calc import SizeCalculator
> +from twisterlib.cmakecache import CMakeCache
>  from twisterlib.handlers import (
>      Handler,
>      SimulationHandler,
> @@ -53,7 +54,6 @@ class TestInstance:
>          self.retries = 0
> 
>          self.name = os.path.join(platform.name, testsuite.name)
> -        self.run_id = self._get_run_id()
>          self.dut = None
>          if testsuite.detailed_test_id:
>              self.build_dir = os.path.join(outdir, platform.name, testsuite.name)
> @@ -62,6 +62,7 @@ class TestInstance:
>              source_dir_rel = testsuite.source_dir_rel.rsplit(os.pardir+os.path.sep, 1)[-1]
>              self.build_dir = os.path.join(outdir, platform.name, source_dir_rel, testsuite.name)
> 
> +        self.run_id = self._get_run_id()
>          self.domains = None
> 
>          self.run = False
> @@ -85,10 +86,22 @@ class TestInstance:
>          """ generate run id from instance unique identifier and a random
>          number"""
> 
> -        hash_object = hashlib.md5(self.name.encode())
> -        random_str = f"{random.getrandbits(64)}".encode()
> -        hash_object.update(random_str)
> -        return hash_object.hexdigest()
> +        cmake_cache_path = os.path.join(self.build_dir, "CMakeCache.txt")
> +        try:
> +            cache = CMakeCache.from_file(cmake_cache_path)
> +            cmake_conf = {}
> +            for k in iter(cache):
> +                cmake_conf[k.name] = k.value
> +            run_id = cmake_conf.get('TC_RUNID')
> +            logger.info(f"cached run id: {run_id}")
> +        except FileNotFoundError:
> +            hash_object = hashlib.md5(self.name.encode())
> +            random_str = f"{random.getrandbits(64)}".encode()
> +            hash_object.update(random_str)
> +            run_id = hash_object.hexdigest()
> +            logger.info(f"new run id: {run_id}")
> +
> +        return run_id
> 
>      def add_missing_case_status(self, status, reason=None):
>          for case in self.testcases:
> diff --git a/subsys/testsuite/ztest/CMakeLists.txt b/subsys/testsuite/ztest/CMakeLists.txt
> index 7817bc2f30b..43520c82dc1 100644
> --- a/subsys/testsuite/ztest/CMakeLists.txt
> +++ b/subsys/testsuite/ztest/CMakeLists.txt
> @@ -11,6 +11,7 @@ zephyr_include_directories(
>    )
> 
>  if(DEFINED TC_RUNID)
> +  set(TC_RUNID ${TC_RUNID} CACHE STRING "testsuite run id")
>    zephyr_compile_definitions(TC_RUNID=${TC_RUNID})
>  endif()
> ```

Your solution will work in terms of incremental builds.
However, it will break current protection mechanisms which is provided by randomly generated run_id for each run for each test.
Example:
running single test on a single platform, run by run, to check code changes for the specific test.
In your proposal, each of the run will have the same run_id, thus there won't be possible to detect if correct code was flashed and executed.
If this case do not have to be supported, then the whole code of generating run_id randomly can be much simpler, it can just use hash of test's name (thus be constant for particular test/platform and allow for incremental build).

At first try I considered solution where run_id was stored in header file and its path was constant (build folder) (thus not need to pass it thought cmake parameter). However, this approach was not working in all cases, as when multiimage/sysbuild tests are build, their cmake build folder is different and this is not easily detectable from twister context.

The idea with storing some information within header file I took from zephyr itself, in this way 'include/generated/version.h` is used to provide information about code version into build system.

If you have some a specific downside of using header file to pass run_id, I will be happy to discuss it ;)

nashif commented:
> running single test on a single platform, run by run, to check code changes for the specific test.
> In your proposal, each of the run will have the same run_id, thus there won't be possible to detect if correct code was flashed and executed.

not sure I am following here. What does 'run by run' means here? Each time you build and run a test with twister it is a new build environment, so we can't have the same run_id. If you build with --no-clean, run_id is reused, so I am not following how the same ID would still be available if you do a clean build.

nordic-piks commented:
> > running single test on a single platform, run by run, to check code changes for the specific test.
> > In your proposal, each of the run will have the same run_id, thus there won't be possible to detect if correct code was flashed and executed.
> 
> not sure I am following here. What does 'run by run' means here? Each time you build and run a test with twister it is a new build environment, so we can't have the same run_id. If you build with --no-clean, run_id is reused, so I am not following how the same ID would still be available if you do a clean build.

Sorry, I was too quick in my explanation.
Let me be more descriptive.
There are 3 solutions:
current - let's call it `run_id_as_parameter`,
proposed in this PR, aka `run_id_in_header`
your proposal, aka `run_id_as_parameter_with_cache`.

I will consider only builds with --no-clean flag, as without this flag there are no incremental builds.

**Calling twister first time:**
```./zephyr/scripts/twister --ninja --inline-logs  --outdir twister-build  --jobs 1 -T zephyr/tests/net/buf --platform nrf52840dk_nrf52840 -v -v --no-clean --device-testing --device-serial /dev/ttyACM0```

`run_id_as_parameter`: twister will generate a random run_id and pass it to cmake call, test builds, flashes and executes on HW, twister is checking if that generated run_id is present at the console output

`run_id_in_header`: twister will generate a random run_id, write it to a file, and pass the path to that file into cmake call, test builds, flashes and executes on HW, twister is checking if that generated run_id is present at the console output

`run_id_as_parameter_with_cache`: in this case the same as option `run_id_as_parameter`

**(Doing small change in test code and) Calling twister second time:**
```./zephyr/scripts/twister --ninja --inline-logs  --outdir twister-build  --jobs 1 -T zephyr/tests/net/buf --platform nrf52840dk_nrf52840 -v -v --no-clean --device-testing --device-serial /dev/ttyACM0```

`run_id_as_parameter`: twister will generate **new random run_id** and pass it to cmake call. All files need to be built from scratch as **cmake parameters have changed - different run id**. Flashing and execution on HW, the twister is checking if that **new generated run_id** is present at the console output

`run_id_in_header`: twister will **generate random run_id**, write it to file, and **pass the same path to file**  into cmake call so that cmake does not see parameter change. Only ninja during build detects that header was changed and rebuilds only files that depend on it - much less than a full build. Flashing and execution on HW, the twister is checking if that **new generated run_id** is present at the console output.

`run_id_as_parameter_with_cache`: twister **will not generate a new random run_id** but it will read it from the cmake cache file, then it will pass it (the same value) to the cmake call. No cmake parameter changed, ninja will build changed files - much less than a full build.  Flashing and execution on HW, the twister is checking if that **previously generated run_id** is present at the console output. This is the behavior change that will be introduced in this solution.
If flashing fails or build fails and there is flashed previous hex file, the run_id value in the console will be still the same, making it impossible to detect such failures.

nashif commented:
> (Doing small change in test code and) Calling twister second time:

thats where the disconnect is, why do you do that? This is not a supported usecase, it assumes incremental builds with code changes are 100% safe, which is not the case.

nordic-piks commented:
> > (Doing small change in test code and) Calling twister second time:
> 
> thats where the disconnect is, why do you do that? This is not a supported usecase, it assumes incremental builds with code changes are 100% safe, which is not the case.

Why do I do incremental builds?
To save a lot of time and resources.

I'm aware that twister does not support it completely, despite available option `--no-clean` with description `Re-use the outdir before building. Will result in faster compilation since builds will be incremental.`

With this PR I want to change twister so that it will not prevent incremental builds, it will only allow them.

I'm aware of potential issues and risks connected to incremental builds, e.g.: dependencies not updated, old code or artifacts used from previous builds.
However, the risk can be mitigated in most cases.
AFAIK, the main problem is connected with cmake not updating its dependencies when there is code/build system/KCONFIG change. To prevent those issues before running tests I remove `CMakeCache.txt` from build folder. This force cmake to regenerate dependencies. (this is a kind of equivalent of cmake --fresh - https://cmake.org/cmake/help/latest/manual/cmake.1.html#cmdoption-cmake-fresh).
I did not noticed any issues with ninja and incremental builds.

Regardless of some cmake issues and how they are solved, incremental builds can coexist with clean builds.
If there is an issue, clean build can be triggered.
In most cases that will give a time saving from incremental build and in rare cases there will be needed clean build (which is now happening every time)

nashif commented:
> I'm aware that twister does not support it completely, despite available option `--no-clean` with description `Re-use the outdir before building. Will result in faster compilation since builds will be incremental.`

This is not about twister, this is a build system issue. The fact we have this in twister is for convinience, mostly for developers, but using this to do validation comes with a lot of risk.

I am not trying to prevent you from doing this, I am just looking for the best, least intrusive way to do that. So, the cmake cache solution does not work because you remove the cache file as you say above? In that case, why not store the in some text file (not a header) and have twister look for such file and reuse the ID if present? Why did you go for a header based approach? very redumentary implementation of this would be:

```diff
index 958019b411a..11fd3fb1453 100644
--- a/scripts/pylib/twister/twisterlib/testinstance.py
+++ b/scripts/pylib/twister/twisterlib/testinstance.py
@@ -53,7 +53,6 @@ class TestInstance:
         self.retries = 0

         self.name = os.path.join(platform.name, testsuite.name)
-        self.run_id = self._get_run_id()
         self.dut = None
         if testsuite.detailed_test_id:
             self.build_dir = os.path.join(outdir, platform.name, testsuite.name)
@@ -62,6 +61,7 @@ class TestInstance:
             source_dir_rel = testsuite.source_dir_rel.rsplit(os.pardir+os.path.sep, 1)[-1]
             self.build_dir = os.path.join(outdir, platform.name, source_dir_rel, testsuite.name)

+        self.run_id = self._get_run_id()
         self.domains = None

         self.run = False
@@ -85,10 +85,22 @@ class TestInstance:
         """ generate run id from instance unique identifier and a random
         number"""

+        run_id_file = os.path.join(self.build_dir, "run_id.txt")
+        if os.path.exists(run_id_file):
+            with open(run_id_file, "r") as fp:
+                run_id = fp.read()
+            if run_id:
+                return run_id
+
         hash_object = hashlib.md5(self.name.encode())
         random_str = f"{random.getrandbits(64)}".encode()
         hash_object.update(random_str)
-        return hash_object.hexdigest()
+        _id = hash_object.hexdigest()
+        os.makedirs(self.build_dir, exist_ok=True)
+        with open(run_id_file, 'w+') as fp:
+            fp.write(_id)
+
+        return _id
```

carlescufi commented:
@nordic-piks can you please rebase?

nordic-piks commented:
> @nordic-piks can you please rebase?

Rebased.

nashif commented:
> That last element is actual problem and regression - that run_id will not change between runs.
> Currently run_id is generated randomly for every run.
> This prevents a lots of failures as twister is checking if newly generated run_id is present at console output.
> Your caching solution will break that mechanisms.

I would not go that far to call this a regression. The run_id is a guard used in the context of one run to deal with issues where you might have some tests getting stuck and and twister recording a PASS on an older test flashed on the device. In optimal cases twister should run fine without the run_id in place.

The stored id will only be resued in the case of --no-clean (which itself has many potential issues on its own), so that is not going to be worse than incremental builds on changed code and tinkering with CMake files and so on.


here some numbers, first with my patch above applied:

**build from scratch with**
`./scripts/twister -p qemu_x86  -T tests/kernel/threads  -v  -x=USE_CCACHE=0`

total build time: 361.68000000000006 sec

**with --no-clean**
`./scripts/twister -p qemu_x86  -T tests/kernel/threads  -v  -x=USE_CCACHE=0 --no-clean`

total build time: 146.21 sec

**with --aggresive-no-clean:**

total build time: 105.03999999999999 sec

**with --aggresive-no-clean and 1 kernel file changed:**

total build time: 140.29 sec

**with ccache only:**

total build time: 276.99999999999994 sec

**with this PR and --no-clean:**

total build time: 157.01999999999998 sec

**with this PR and --no-clean and cache files removed**

total build time: 171.81 sec

nordic-piks commented:
> > That last element is actual problem and regression - that run_id will not change between runs.
> > Currently run_id is generated randomly for every run.
> > This prevents a lots of failures as twister is checking if newly generated run_id is present at console output.
> > Your caching solution will break that mechanisms.
> 
> I would not go that far to call this a regression. The run_id is a guard used in the context of one run to deal with issues where you might have some tests getting stuck and and twister recording a PASS on an older test flashed on the device. In optimal cases twister should run fine without the run_id in place.
> 
> The stored id will only be resued in the case of --no-clean (which itself has many potential issues on its own), so that is not going to be worse than incremental builds on changed code and tinkering with CMake files and so on.
> 
> here some numbers, first with my patch above applied:
> 
> **build from scratch with** `./scripts/twister -p qemu_x86 -T tests/kernel/threads -v -x=USE_CCACHE=0`
> 
> total build time: 361.68000000000006 sec
> 
> **with --no-clean** `./scripts/twister -p qemu_x86 -T tests/kernel/threads -v -x=USE_CCACHE=0 --no-clean`
> 
> total build time: 146.21 sec
> 
> **with --aggresive-no-clean:**
> 
> total build time: 105.03999999999999 sec
> 
> **with --aggresive-no-clean and 1 kernel file changed:**
> 
> total build time: 140.29 sec
> 
> **with ccache only:**
> 
> total build time: 276.99999999999994 sec
> 
> **with this PR and --no-clean:**
> 
> total build time: 157.01999999999998 sec
> 
> **with this PR and --no-clean and cache files removed**
> 
> total build time: 171.81 sec

Let's go to conclusions and future of this PR.
My current focus is to have working incremental builds, regardless of the way how it is implemented at twister
(of course, without breaking other features).
Thus, can we agree that implementation can look like what you have already proposed here:
https://github.com/zephyrproject-rtos/zephyr/pull/65003#issuecomment-1821678983
(I modified it slightly - run id value is stored in dedicated file at build folder)
```
index d3cbbaf498..6635865827 100644
--- a/scripts/pylib/twister/twisterlib/testinstance.py
+++ b/scripts/pylib/twister/twisterlib/testinstance.py
@@ -54,7 +54,6 @@ class TestInstance:
         self.retries = 0

         self.name = os.path.join(platform.name, testsuite.name)
-        self.run_id = self._get_run_id()
         self.dut = None
         if testsuite.detailed_test_id:
             self.build_dir = os.path.join(outdir, platform.name, testsuite.name)
@@ -63,6 +62,7 @@ class TestInstance:
             source_dir_rel = testsuite.source_dir_rel.rsplit(os.pardir+os.path.sep, 1)[-1]
             self.build_dir = os.path.join(outdir, platform.name, source_dir_rel, testsuite.name)

+        self.run_id = self._get_run_id()
         self.domains = None

         self.run = False
@@ -84,12 +84,22 @@ class TestInstance:

     def _get_run_id(self):
         """ generate run id from instance unique identifier and a random
-        number"""
-
-        hash_object = hashlib.md5(self.name.encode())
-        random_str = f"{random.getrandbits(64)}".encode()
-        hash_object.update(random_str)
-        return hash_object.hexdigest()
+        number
+        If exist, run cached run id from previous run."""
+        run_id = ""
+        run_id_file = os.path.join(self.build_dir, "run_id.txt")
+        if os.path.exists(run_id_file):
+            with open(run_id_file, "r") as fp:
+                run_id = fp.read()
+        else:
+            hash_object = hashlib.md5(self.name.encode())
+            random_str = f"{random.getrandbits(64)}".encode()
+            hash_object.update(random_str)
+            run_id = hash_object.hexdigest()
+            os.makedirs(self.build_dir, exist_ok=True)
+            with open(run_id_file, 'w+') as fp:
+                fp.write(run_id)
+        return run_id

     def add_missing_case_status(self, status, reason=None):
         for case in self.testcases:
```

What do you think?
Should I do another PR or modify this one?

I you already measured, this reduces build time ~50%.
From my perspective, --no-clean can be used at CI when adding before twister execution additional command:
`find twister-build \( -name '*.log' -o -name 'CMakeCache.txt' -o -wholename '*/tmp/*' -o -wholename '*zephyr/misc/generated/*' -o -wholename '*-stamp/*'  -o -name '.config' \) -delete`
It will remove some output file, so that it will force rebuild of needed staff (that is currently working for me, possible this line may evolve).

For me, `--aggresive-no-clean` cannot be used at CI, not running cmake will lead to errors when dependencies has changed.
Currently, removing `CMakeCache.txt` and `.config` solves that issues but running cmake is needed.

nashif commented:
> What do you think?
> Should I do another PR or modify this one?

sounds good, please just reuse this PR

nordic-piks committed:
scripts: twister: testinstance: store run id between builds

Cache value of run id between builds.
Save it in file and later load during next build.

Signed-off-by: Piotr Kosycarz <piotr.kosycarz@nordicsemi.no>

nordic-piks commented:
Just updated docstring for method and removed accidentally added test's output files.
